Backends/CLX/clipboard.lisp

```diff
@@ -0,0 +1,285 @@
+  (let ((sheet (selection-object-owner requested-object))
+        (window (sheet-xmirror (selection-object-owner requested-object)))
+        (to-type (selection-object-type requested-object))
+        (table (climi::get-object-table requested-object)))
+    (labels ((wait-for-request ()
+               (or (not (clipboard-outstanding-request port))
+                   (not (eq selection (selection-object-content requested-object)))))
+             (negotiate-selection-type ()
+               (xlib:convert-selection selection :targets window :mcclim nil)
+               (event-listen-or-wait sheet :wait-function #'wait-for-request :timeout 1)
+               ;; Either window is closed or peer does not understand
+               ;; targets. We will try to request string in that case.
+               (unless (clipboard-outstanding-request port)
+                 (setf (clipboard-outstanding-request port) requested-object
+                       (selection-object-content requested-object) '(:utf8_string)))
+               ;; We look for the first proposed type which is suitable for the
+               ;; selection translation. If not found it is NIL.
+               ;;
+               ;; XXX: we should collect all applicable translators and their
+               ;; types and sort them by priority as they are sorted in
+               ;; find-presentation-translators. -- jd 2019-06-14
+               (setf (selection-object-type requested-object) nil)
+               (some (lambda (from-type)
+                       (when (or (and (not (eql to-type t))
+                                      (subtypep from-type to-type))
+                                 (find-selection-translators from-type to-type table))
+                         (setf (selection-object-type requested-object) from-type)))
+                     (selection-object-content requested-object))))
+      (declare (dynamic-extent #'wait-for-request))
+      (negotiate-selection-type)
+      ;; rewind the content to selection
+      (when-let ((from-type (selection-object-type requested-object)))
+        (setf (selection-object-content requested-object) selection)
+        (xlib:convert-selection selection from-type window :mcclim nil)
+        (when (event-listen-or-wait sheet :wait-function #'wait-for-request :timeout 1)
+          (when-let ((result (clipboard-outstanding-request port)))
+            (call-selection-translator (selection-object-content result)
+                                       (selection-object-type result)
+                                       to-type
+                                       table)))))))
+
+(macrolet ((frob (selection)
+             `(defmethod request-selection ((port clx-selection-mixin)
+                                            (selection (eql ,selection))
+                                            (requested-object selection-object))
+                (when (climi::stored-object port selection)
```

Comment:
This check sometimes succeeds when the selection isn't actually owned. More details here: https://github.com/McCLIM/McCLIM/pull/775#discussion_r293679246

Reply:
What seems to happen is that if the selection ownership is released when the window is closed, the ownership release event is never delivered. Thus the local selection will not be reset until it acquires the selection again, after which the next selection will properly deliver a release ever.

Reply:
fix pushed